### DISCUSSION TITLE
Recommendation to encode (lists of) flat key-value pairs

### DISCUSSION THREAD

**Discussion by nichtich** [kdl-org/kdl#150] [Q&A] [answered]
Not having internalized the data model of KDL yet, I [struggled](https://github.com/kdl-org/kdl/issues/146) to see how to encode a flat set of key-value pairs. Given this example in INI File syntax:

~~~ini
answer=42
author="Douglas Adams"
work="The Hitchhiker's Guide to the Galaxy"
~~~

The KDL approach could be either

~~~kdl
answer 42
author "Douglas Adams"
work "The Hitchhiker's Guide to the Galaxy"
~~~

or

~~~ini
- answer=42 author="Douglas Adams" work="The Hitchhiker's Guide to the Galaxy"
~~~

The latter looks more familiar for readers of other data structuring languages but it is more difficult to edit (linebreak) and requires a root node identifier (here: `-`).

There should be a recommendation that also includes lists of flat key-value pairs (e.g. rows from a SQL database or from a CSV file with header), maybe **KViK (key-values in KDL)**. How about recommending these to ways (also mixable):

~~~kdl
- { 
  author "Douglas Adams"
  title "The Hitchhiker's Guide to the Galaxy"
}
- { 
  author "Ursula K. Le Guin"
  title "A Wizard of Earthsea"
}
~~~

~~~kdl
- author="Douglas Adams"  title="The Hitchhiker's Guide to the Galaxy"
- author="Ursula K. Le Guin"  title="A Wizard of Earthsea"
~~~

So each node must have node name `-` and *either* no arguments nor children block *or* a children block but no arguments nor parameters and nodes in the children block must have distinct names and exactly one argument.

Encoding [INI Files](https://en.wikipedia.org/wiki/INI_file) is straightforward if you allow other node names in addition to `-` as section names and allow nodes with one argument for the  global section.

~~~
name value # global section

section a=1 b=2

another-section {
  a 1
  b 2
}

- a=1 b=2  # can be read as row
~~~

I'm sure such a limited set of KDL, compatible with INI files and (rows of) key-value pairs would be useful.

Tabular data without header is more difficult but maybe not needed to be specified anyway. Maybe like this:

~~~
matrix rows=3 \
A1 B1 C1 \
A2 B2 C2 \
A3 B3 C3
~~~

**Comment by zkat** [comment, marked as answer]
```kdl
answer 42
author "Douglas Adams"
work "The Hitchhiker's Guide to the Galaxy"
```

This is definitely how I would encode it myself. Properties aren't meant for arbitrary amounts of metadata, but more for limited "metadata" for a node itself. So, a small list of flags, etc. If you want a _lot_ of things, then children is usually the way to go for sure.
